Do primary filtering for the persons endpoint in ClickHouse

Twixes opened the pull request:
## Changes

A proposal aiming to resolve https://github.com/PostHog/posthog/issues/5717.

## Checklist

- [x] All querysets/queries filter by Organization, by Team, and by User
- [x] Django backend tests

Twixes committed:
Do primary filtering for the persons endpoint in ClickHouse

macobo commented on ee/clickhouse/views/person.py:
```diff
@@ -1,30 +1,93 @@
-class ClickhousePersonViewSet(PersonViewSet):
+    if and_conditions:
+        final_query = f"""{BASE_FILTER_PERSONS_QUERY} {' '.join(and_conditions)}"""
+        uuids_found_rows = cast(list, sync_execute(final_query, params))
```
Q: Why do we need the cast here? Seems wrong.

Twixes replied:
As in some other case we discussed in a different PR, Pylance sees types differently than Mypy (i.e. in this case it's afraid it may be a `None`), but can remove :)

macobo commented on ee/clickhouse/views/person.py:
```diff
@@ -1,30 +1,93 @@
+        and_conditions.append(
+            f"""
+        AND (
+            person.properties ILIKE %(icontains)s
```
This is really really sketchy. 

1. We should only match on certain parsed properties not an opaque json blob - the edge cases introduced are kind of insane.
2. This won't be able to make use of materialized columns either
3. This won't match postgres' behavior

Twixes replied:
You know, I agree! But this _does_ match Postgres's behavior, which is the point. `filter_persons()`, which this is an analogue of, literally does `properties__icontains` – and this is reflected in the SQL by `UPPER(posthog_person.properties::string) LIKE UPPER(test_string)` (`UPPER` instead of `ILIKE` I presume because the former is [a _bit_ faster in Postgres](https://www.peterbe.com/plog/upper-vs-ilike).

macobo replied:
Ah! I didn't check out the postgres implementation - I assumed it was based off of email property. :sweat_smile: 

Do the two behave exactly identically though?

I do think this is worth visiting from a product standpoint though before committing to this on clickhouse.

Twixes replied:
They are indeed exactly identical, and actually this is _better_ in ClickHouse, because `properties` is already a UTF-8 string in ClickHouse, while the Postgres version had to (and still does) cast JSONB to a string, which is not free.
I would argue to keep this as is, simply because this is so ridiculously cheap compute-wise, while covering a case that would be really difficult to handle otherwise: search among keys and values of _nested_ properties. I'm okay with this being slightly quirky as a tradeoff.

macobo commented on ee/clickhouse/views/person.py:
```diff
@@ -1,30 +1,93 @@
+WHERE team_id = %(team_id)s"""
+
+
+def filter_persons_ch(team_id: int, request: Request, queryset: QuerySet) -> QuerySet:
```
Doing all of this in the view is really sketchy.

Given that this endpoint exists in both clickhouse and postgres, the proper approach is to extract a query object and have a test_factory testing both.

Twixes replied:
What do you mean by `extract a query object and have a test_factory testing both`?

macobo replied:
In other words it should behave as e.g. trends does. There's an endpoint which creates a Query object and all the work is done in there. There's also a test_factory which creates tests for the feature and is used for both postgres and ee tests.

Twixes replied:
I still don't quite get what this'd look like in a cleaner way.

macobo replied:
See refactoring done in https://github.com/PostHog/posthog/pull/5842 - I separated out a query class, same way e.g. trends and everything else works.

macobo commented on ee/clickhouse/views/person.py:
```diff
@@ -1,30 +1,93 @@
+from posthog.exceptions import UnsupportedFeature
 from posthog.models import Event, Filter, Person
 
+BASE_FILTER_PERSONS_QUERY = """
```
This whole query is wrong:

1. It doesn't account for deleted persons or deleted distinct ids
2. It doesn't do data rollups as our other queries do, needed for https://github.com/PostHog/product-internal/issues/134

Twixes replied:
1. :+1: Will add a non-deleted condition, thanks.
2. What's a data rollup in this context?

macobo replied:
Good question. So you should use ee/clickhouse/sql/person.py#GET_LATEST_PERSON_SQL as a basis rather than rolling your own here. :)

Twixes replied:
Thanks! Didn't have context on this, now reworked the query (I ended up adding `GET_LATEST_PERSON_WITH_DISTINCT_IDS_SQL` to `person.py` `sql/` file, to facilitate search on distinct IDs).

macobo commented on ee/clickhouse/views/test/test_clickhouse_person.py:
```diff
@@ -13,14 +15,38 @@ def _create_event(**kwargs):
 def _create_person(**kwargs):
-    return Person.objects.create(**kwargs)
+    postgres_person: Person = Person.objects.create(**kwargs)
+    sync_execute(
```
This should not be in here and will result in duplicate data.

There are already hooks defined in ee/clickhouse/models/person.py which create the user

In general trying to do inserts into clickhouse from python code is a failing strategy we should avoid - they will cause bugs as the python and plugin server behavior diverge. Example: https://github.com/PostHog/posthog/issues/5145

Twixes replied:
I agree, though persons do need to be in the DB at the time of this test suite running, so not sure what cleaner solution there is. Sending events to the plugin server to create the persons? That's really asynchronous though, we'd have to `sleep` until the persons appear in CH here (which we do routinely do in plugin server tests, so maybe?).

macobo replied:
Clickhouse inserts are synchronous in tests, check out ee/kafka_client/client.py#ClickhouseProducer implementation

macobo commented on ee/clickhouse/views/person.py:
```diff
@@ -1,30 +1,93 @@
+    if request.GET.get("search"):
+        parts = request.GET["search"].split(" ")
+        contains = []
+        for part_index, part in enumerate(parts):
```
Worth extracting a helper to retain identical behavior with postgres' implementation?

Twixes replied:
I feel like a helper would just be more confusing considering that the general logic is the same, but it's the bottom of each branch that looks very different between the two versions.

macobo replied:
If this is causing issues for you, worth making a change to e.g. `def sync_execute` instead in a separate PR? Or using mypy? :)

Twixes replied:
Comment in the wrong place I think (should be in "Q: Why do we need the cast here?")?

macobo replied:
Let's take a step back.

Fundamentally this function is just doing too many things:

1. Extracting "meaning" from search query (e.g. a filter)
  2. Including some very specific parsing logic
2. Constructing query based on that
3. Lastly querying both ch and postgres based off of it

Everything we can do to extract these different pieces fundamentally makes the code easier to follow since it can make use of shared patterns.

Twixes committed:
Remove `sync_execute` from `_create_person`

Django hooks _should_ do the job.

Twixes committed:
Merge branch 'master' into persons-api-clickhouse-primary-filtering

Twixes committed:
Debug

`test_properties` passes locally but there's a server error in CI for some reason.

Twixes committed:
Improve filtering

Twixes committed:
Fix redundant `format` argument

Twixes committed:
Optimize `filter_persons_ch`

Twixes committed:
Debug

Twixes committed:
Disable use of materialized columns in person search

macobo commented on ee/clickhouse/sql/person.py:
```diff
@@ -89,6 +89,33 @@
 
+# Using any() to make GROUP BY cheaper
+# See https://clickhouse.tech/docs/en/sql-reference/aggregate-functions/reference/any/
+GET_LATEST_PERSON_WITH_DISTINCT_IDS_SQL = """
```
Note: This query is 100% this view specific and _should not_ be used anywhere else.

Problem is that GET_LATEST_PERSON_SQL needs to die ASAP - it doesn't allow for querying materialized columns. The proper replacement is `ClickhousePersonQuery` which can account for this.

However since this endpoint is doing the _wrong thing_ :tm: fundamentally around filtering properties this works for now.

Given that this kind of is one query specific hack, let's:
1. Refactor out the query from within the endpoint (there's an existing open thread about this)
2. Move the const there to signify this is that-usage specific
3. Add a comment to that effect.

macobo commented on ee/clickhouse/views/person.py:
```diff
@@ -1,30 +1,94 @@
+        )
+        uuids_found_rows = cast(list, sync_execute(final_query, params))
+        uuids_found = [row[0] for row in uuids_found_rows]
+        queryset = queryset.filter(uuid__in=uuids_found)
```
This smells like a hack.

Why not use `ClickhousePersonSerializer` like other endpoints? Going to postgres in the end is kind of fishy without a good cause.

macobo commented on ee/clickhouse/views/person.py:
```diff
@@ -1,30 +1,94 @@
+
+    if and_conditions:
+        final_query = GET_LATEST_PERSON_WITH_DISTINCT_IDS_SQL.format(
+            query=" ".join(and_conditions).replace("AND ", "", 1)
```
No other code does this and the replacing first and is kind of :TRICKY:

Instead do `query = f"1 = 1 {' '.join(and_conditions)}"` if you're sure all are prefixed with AND (which is non-obvious).

macobo reviewed (changes requested):
Added some comments.

Let's leave the properties ILIKE matching in here for now, but other feedback here is IMO kind of blocking. Let's jump on a call if anything is unclear!
